src/org/sosy_lab/java_smt/basicimpl/Tokenizer.java

```diff
@@ -0,0 +1,202 @@
+            if (c == '(') {
+              // Handle opening brackets
+              token.append("(");
+              level++;
+            } else {
+              // Should be unreachable: all top-level expressions need parentheses around them
+              throw new IllegalArgumentException();
+            }
+          }
+        } else {
+          // We're inside an r-expression
+          token.append(c);
+          // Handle opening/closing brackets
+          if (c == '(') {
+            level++;
+          }
+          if (c == ')') {
+            if (level == 1) {
+              builder.add(token.toString());
+              token = new StringBuilder();
+            }
+            level--;
+          }
+        }
+      }
+    }
+    if (level != 0) {
+      // Throw an exception if the brackets don't match
+      throw new IllegalArgumentException();
+    }
+    return builder.build();
+  }
+
+  private static boolean matchesOneOf(String token, String... regexp) {
+    return token.matches("\\(\\s*(" + String.join("|", regexp) + ")[\\S\\s]*");
+  }
+
+  /**
+   * Check if the token is a function or variable declaration.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  public static boolean isDeclarationToken(String token) {
+    return matchesOneOf(token, "declare-const", "declare-fun");
+  }
+
+  /**
+   * Check if the token is a function definition.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  public static boolean isDefinitionToken(String token) {
+    return matchesOneOf(token, "define-fun");
+  }
+
+  /**
+   * Check if the token is an <code>(assert ...)</code>.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  public static boolean isAssertToken(String token) {
+    return matchesOneOf(token, "assert");
+  }
+
+  /**
+   * Check if the token is <code>(set-logic ..)</code>.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  public static boolean isSetLogicToken(String token) {
+    return matchesOneOf(token, "set-logic");
+  }
+
+  /**
+   * Check if the token is <code>(exit ...)</code>.
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  public static boolean isExitToken(String token) {
+    return matchesOneOf(token, "exit");
+  }
+
+  /**
+   * Check if this is an unsupported token.
+   *
+   * <p>The list of unsupported tokens contains:
+   *
+   * <ul>
+   *   <li>push
+   *   <li>pop
+   *   <li>reset-assertions
+   *   <li>reset
+   * </ul>
+   *
+   * <p>Use {@link #tokenize(String)} to turn an SMT-LIB2 script into a string of input tokens.
+   */
+  public static boolean isUnsupportedToken(String token) {
+    return matchesOneOf(token, "push", "pop", "reset-assertions", "reset");
```

Comment:
Instead of using a list of unsupported commands, it might be simpler to list the supported ones.
--> return !matchesOneOf(token, "assert", "exit", ...)

Reply:
There are some commands that are neither `supported`, nor `unsupported`. An example is `(check-sat)` which Bitwuzla will append to any SMTLIB formula it dumps. These commands should be skipped by the sanitizer as they are not needed. Throwing an exception would cause issues with some solvers.

I've renamed `isUnsupportedToken` to `isForbiddenToken` to make the difference more clear.